### PR TITLE
general tidy up in anticipation of RFC

### PR DESCRIPTION
**Pull request by charliemirabile**
- updated framebuffer test to avoid mmap since it will be going away
- added a new test that checks the function of each bit in the pixels
- minor upgrade to makefile unload rule so it doesn't report and error the modules are already unloaded
- combined the three headers (`core.h`, `joystick.h`, and `framebuffer.h`) into one new header `rpisense.h`
- general tidy up of `rpisense-js.c` and `rpisense-core.c`
- new error checks in core probe
- new interface between core and platform drivers that abstracts the details of the i2c communications

### CODE REVIEW

File: rpisense-core.c [hunk 1]
```diff
@@ -20,37 +20,24 @@
...
-static void rpisense_client_dev_register(struct rpisense *rpisense,
-					 const char *name,
-					 struct platform_device **pdev)
-{
-	int ret;
+#define RPISENSE_FB			0x00
+#define RPISENSE_WAI			0xF0
+#define RPISENSE_VER			0xF1
+#define RPISENSE_KEYS			0xF2
+#define RPISENSE_EE_WP			0xF3
```
Comment (theyoyojo) [inline rpisense-core.c:31]: what is the reasoning moving these from the header file into core.c?

File: rpisense-core.c [hunk 2]
```diff
@@ -60,56 +47,75 @@ static int rpisense_probe(struct i2c_client *i2c,
...
-}
+	pdev->dev.parent = rpisense->dev;
+	platform_set_drvdata(pdev, rpisense);
 
+	ret = platform_device_add(pdev);
+	if (ret != 0)
+		goto add_fail;
 
-s32 rpisense_reg_read(struct rpisense *rpisense, int reg)
+	ret = devm_add_action_or_reset(rpisense->dev,
```
Comment (theyoyojo) [inline rpisense-core.c:99]: nice all of this is way cleaner, removed a level of indentation too

File: rpisense-core.c [hunk 3]
```diff
@@ -60,56 +47,75 @@ static int rpisense_probe(struct i2c_client *i2c,
...
+
+	return pdev;
+
+add_fail:
+	platform_device_put(pdev);
+alloc_fail:
+	return ERR_PTR(ret);
+}
+
+int rpisense_get_joystick_state(struct rpisense *rpisense)
```
Comment (theyoyojo) [inline rpisense-core.c:112]: do you think this might better belong in rpisense-js?

File: rpisense-core.c [hunk 4]
```diff
@@ -60,56 +47,75 @@ static int rpisense_probe(struct i2c_client *i2c,
...
-	int ret = i2c_smbus_read_byte_data(rpisense->i2c_client, reg);
+	int ret = i2c_smbus_read_byte_data(rpisense->i2c_client, RPISENSE_KEYS);
 
-	if (ret < 0)
-		dev_err(rpisense->dev, "Read from reg %d failed\n", reg);
-	/* Due to the BCM283x I2C clock stretching bug, some values
-	 * may have MSB set. Clear it to avoid incorrect values.
-	 * */
-	return ret & 0x7F;
+	return ret < 0 ? ret : ret & 0x1f;
```
Comment (theyoyojo) [inline rpisense-core.c:116]: why 0x1f vs 0xx7F?

File: rpisense-core.c [hunk 5]
```diff
@@ -20,37 +20,24 @@
...
-	platform_set_drvdata(*pdev, rpisense);
-	ret = platform_device_add(*pdev);
-	if (ret != 0) {
-		dev_err(rpisense->dev, "Failed to register %s: %d\n",
-			name, ret);
-		platform_device_put(*pdev);
-		*pdev = NULL;
-	}
-}
+static struct platform_device *
```
Comment (theyoyojo) [inline rpisense-core.c:34]: kind of stands out to have this lone forward decl when we could just move the function to this location of the code

File: rpisense-js.c [hunk 6]
```diff
@@ -14,30 +14,31 @@
...
 	struct rpisense_js *rpisense_js = &rpisense->joystick;
-	s32 keys = rpisense_reg_read(rpisense, RPISENSE_KEYS);
+	s32 keys = rpisense_get_joystick_state(rpisense);
 	s32 changes = keys ^ prev_keys;
 
 	prev_keys = keys;
-	for (i = 0; i < ARRAY_SIZE(keymap); i++) {
-		if (changes & 1) {
+	for (i = 0; i < ARRAY_SIZE(keymap); ++i) {
+		if (changes & (1<<i)) {
```
Comment (theyoyojo) [inline rpisense-js.c:38]: nice

File: rpisense-js.c [hunk 7]
```diff
@@ -47,11 +48,19 @@ static int rpisense_js_probe(struct platform_device *pdev)
...
-	if (!rpisense_js->keys_dev) {
+	rpisense_js->keys_desc = devm_gpiod_get(&rpisense->i2c_client->dev,
+						"keys-int", GPIOD_IN);
+	if (IS_ERR(rpisense_js->keys_desc)) {
+		dev_warn(&pdev->dev, "Failed to get keys-int descriptor.\n");
+		return PTR_ERR(rpisense_js->keys_desc);
+	}
+
+
+	rpisense_js->keys_dev = devm_input_allocate_device(&pdev->dev);
```
Comment (theyoyojo) [inline rpisense-js.c:62]: nice

**Review by theyoyojo — CHANGES_REQUESTED** [review]
looks good, just a few questions about a couple of things

File: rpisense-core.c (hunk 1 above)
Comment (charliemirabile) [reply rpisense-core.c:31]: wanted to hide the details of i2c from the public header. This change makes more sense in context of what I did for the frame buffer. I actually moved a lot of the code for refreshing the display  into the function that replaces block write.

File: rpisense-core.c (hunk 3 above)
Comment (charliemirabile) [reply rpisense-core.c:112]: see reasoning above for why I moved the constants into core

File: rpisense-core.c (hunk 4 above)
Comment (charliemirabile) [reply rpisense-core.c:116]: joystick only has 5 buttons so I just mask out only the relevant portion. Up to debate whether that is worth doing

File: rpisense-core.c (hunk 5 above)
Comment (charliemirabile) [reply rpisense-core.c:34]: I wanted probe and remove at the top of each file. When I rewrote frame buffer I also had to add a single forward decl. to make that possible. Obviously this is a matter of taste, but I think I prefer probe and remove at top as opposed to whatever other random crap. makes it easier to follow the flow of code

File: rpisense-core.c (hunk 4 above)
Comment (theyoyojo) [reply rpisense-core.c:116]: Ok that's fair

**Review by theyoyojo — APPROVED** [review]

**Pull request merged**